### DISCUSSION TITLE
Can we set a variable initial atom value and have it available on first render?

### DISCUSSION THREAD

**Discussion by grazianodev** [pmndrs/jotai#1362] [Q&A]
Hi, 

I have a use case that I am struggling with and not sure how to achieve. Sorry if it's a silly question, I'm not an experienced developer :)

Suppose I have the below Nextjs app. Right now, `type` always resolves to "video" of course, because that's the initial value set for `typeAtom`. The question is, how can I set a different initial value based on which page I am in? 

I know I could set the type in `useEffect` and set the new value there, but then I would have to create a loading state for that first render... Is there a way to change the value and have it available on first render already? I've tried adding a Provider with `initialValues` in each page, but then the `RenameBookmarksCollection` component does not inherit that value because it's not within that Provider (and I can't move it within it!)

```
// _app.js

const MyApp = ( { Component, pageProps } ) =>
   <div className="App">
      <Provider>
         <Component { ...pageProps } />
         <RenameBookmarksCollection />
      </Provider>
   </div>   

// atoms.js

export const typeAtom = atom( "video" )

// videos.js

const VideoBookmarksPage = () => {
   const [ type ] = useAtom( typeAtom )
   return(
      <div className="Page">
         <p>Bookmarks of type: { type }</p>
      </div>
   )
}

// images.js

const ImageBookmarksPage = () => {
   const [ type ] = useAtom( typeAtom )
   return(
      <div className="Page">
         <p>Bookmarks of type: { type }</p>
      </div>
   )
}

// RenameBookmarksCollection.js

const RenameBookmarksCollection = () => {
   const [ type ] = useAtom( typeAtom )
   return(
      <div className="View">
         <p>Rename collection of bookmarked { type }</p>
      </div>
   )
}

```

Thanks for any help!

**Comment by dai-shi** [comment]
`initialValues` should be the answer, but

> then the `RenameBookmarksCollection` component does not inherit that value because it's not within that Provider (and I can't move it within it!)

what does that mean? `RenameBookmarksCollection` seems to be inside the Provider.

**Comment by grazianodev** [comment]
Sorry if that wasn't clear. I mean that, if I had a Provider with initialValues in each page (not in _app.js), then `RenameBookmarksCollection` does not inherit the type. For example, if I do this: 

```
const Page = () => {
   const [ type ] = useAtom( typeAtom )
   return(
      <div className="Page">
         <p>Bookmarks of type: { type }</p>
      </div>
   )
}

const ImageBookmarksPage = () => 
   <Provider initialValues={[[typeAtom,"images"]]}>
      <Page />
   </Provider>
```

then the type is correctly changed to "images" in the page, but it's still "video" in `RenameBookmarksCollection`. I'm not getting the value of `type` from the server, my idea was simply to hardcode it in each bookmarks page 'cause I will only have a few.

If there isn't another way, I guess I could set the type in each page's `getStaticProps` so that I can retrieve it in _app.js and inject it in the initialValues of the root Provider.

**Reply by dai-shi** [reply]
Oh, I misunderstood because you have `<Provider>` in `_app.js`.
So, do you mean you can't add `initialValues={[[typeAtom,"images"]]}` in your `<Provider>` in `_app.js`? Where do you get `"images"`?

**Reply by grazianodev** [reply]
I was going to hardcode "images" because I only have a few different types. So, I could add `initialValues` in the root Provider in _app, but I would have to pass the initial value from each page via getStaticProps (which I don't need right now): 

```
// _app.js

const MyApp = ( { Component, pageProps } ) => 
   <div className="App">
      <Provider initialValues={[[typeAtom,pageProps.type]]}>
         <Component { ...pageProps } />
         <RenameBookmarksCollection />
      </Provider>
   </div> 

// images.js

const ImageBookmarksPage = ( { type } ) => {
   return(
      <div className="Page">
         <p>Bookmarks of type: { type }</p>
      </div>
   )
}

export const getStaticProps = () => {
   return {
      props: {
         type: "images"
      }
   }
}
```

The above works, and that's how I'll do it if there isn't another way :)

**Reply by dai-shi** [reply]
Yes, the `initialValues` prop in `Provider` should be the first and stable method.

> but I would have to pass the initial value from each page via getStaticProps (which I don't need right now)

In such a case, `useHydrateAtoms` should help, but it's rather unstable/experimental.

**Reply by grazianodev** [reply]
Yes, I had tried using `useHydrateAtoms` in each page but the value is only set once and not updated (as stated in the docs). That would be nice if it updated without having to remount the provider. In the meanwhile, I'll use the above solution. Thank you for your help!

**Reply by dai-shi** [reply]
To update after mount, `useEffect` is a only valid solution. One could create a custom hook combining `useHydrateAtoms` and `useEffect`(+`useSetAtom`).
